n2y/converter.py

```diff
@@ -0,0 +1,466 @@
+    TableFoot, RowHeadColumns, Row, Cell, RowSpan, ColSpan, ColWidthDefault, AlignDefault, Caption
+import re
+
+from n2y.notion import Client
+
+
+# Notes:
+# A single Notion block may have multiple lines of text.
+# A page is a block that puts children into "content" attribute.
+# We transform page block to resemble other block types.
+#
+# Pandoc makes each word a block, and spaces are blocks too!
+#
+# Block types used here that do not exist in Notion:
+#   container - block with no top-level content, only children (used to parse a page and lists)
+#   bulleted_list - Notion has bulleted_list_item, but no enclosing container
+#   numbered_list - Notion has numbered_list_item, but no enclusing container
+
+
+IMAGE_PATH = None
+IMAGE_WEB_PATH = None
+
+
+def load_plugins(filename):
+    global Bookmark, BulletedList, BulletedListItem, ChildPageBlock, CodeBlockFenced, \
+        Divider, HeadingOne, HeadingTwo, HeadingThree, ImageBlock, NumberedList, NumberedListItem, \
+        ParagraphBlock, Quote, RowBlock, TableBlock, ToDo, ToDoItem, Toggle
+    abs_path = path.abspath(filename)
+    plugin_spec = importlib.util.spec_from_file_location("plugins", abs_path)
+    plugin_module = importlib.util.module_from_spec(plugin_spec)
+    plugin_spec.loader.exec_module(plugin_module)
+    print(plugin_module)
+    for (key, value) in plugin_module.exports.items():
+        if key == "Bookmark":
+            Bookmark = value
+        elif key == "BulletedList":
+            BulletedList = value
+        elif key == "BulletedListItem":
+            BulletedListItem = value
+        elif key == "ChildPageBlock":
+            ChildPageBlock = value
+        elif key == "CodeBlockFenced":
+            CodeBlockFenced = value
+        elif key == "Divider":
+            Divider = value
+        elif key == "HeadingOne":
+            HeadingOne = value
+        elif key == "HeadingTwo":
+            HeadingTwo = value
+        elif key == "HeadingThree":
+            HeadingThree = value
+        elif key == "ImageBlock":
+            ImageBlock = value
+        elif key == "NumberedList":
+            NumberedList = value
+        elif key == "NumberedListItem":
+            NumberedListItem = value
+        elif key == "ParagraphBlock":
+            ParagraphBlock = value
+        elif key == "Quote":
+            Quote = value
+        elif key == "RowBlock":
+            RowBlock = value
+        elif key == "TableBlock":
+            TableBlock = value
+        elif key == "ToDo":
+            ToDo = value
+        elif key == "ToDoItem":
+            ToDoItem = value
+        elif key == "Toggle":
+            Toggle = value
+        else:
+            raise NotImplementedError(f"Unknown plugin type {key}")
+
+
+def load_block(client: Client, id, get_children=True):
+    block = client.get_block(id)
+    return parse_block(client, block, get_children)
+
+
+def parse_block(client: Client, block, get_children=True):
```

Comment:
It feels a little weird to me that we're passing the notion client down into all of the blocks. I may be off base with this, but it feels like it would be nice if we could conceptually divide the code into two discrete stages:

1. Grabbing data from Notion
2. Converting this data into our own format

Separating it out into two stages like this would remove the need for patching the client in the tests.

I guess one advantage of intermixing them is that users could write a plugin that decides to skip over certain blocks that it's not interested in, thus saving API calls.

What do you think? Would it be possible to split up the work into two discrete stages, or is there some reason why intermixing the data-fetching with the data conversion is better.

Reply:
That makes sense! If you haven't already, I think it would be worth copying your response to my GitHub comment into a code comment somewhere. It seems like the sort of high-level "why" question that's unlikely to go stale and may help future people working on the project understand the architecture.